src/gc-alloc-profiler.cpp

```diff
@@ -49,15 +49,22 @@ jl_combined_results g_combined_results; // Will live forever.
 // === stack stuff ===
 
 jl_raw_backtrace_t get_raw_backtrace() JL_NOTSAFEPOINT {
-    // A single large buffer to record backtraces onto
-    static jl_bt_element_t static_bt_data[JL_MAX_BT_SIZE];
+    // We first record the backtrace onto a MAX-sized buffer, so that we don't have to
+    // allocate the buffer until we know the size. To ensure thread-safety, we use a
+    // per-thread backtrace buffer.
+    jl_ptls_t ptls = jl_current_task->ptls;
+    jl_bt_element_t *shared_bt_data_buffer = ptls->profiling_bt_buffer;
 
-    size_t bt_size = rec_backtrace(static_bt_data, JL_MAX_BT_SIZE, 2);
+    size_t bt_size = rec_backtrace(shared_bt_data_buffer, JL_MAX_BT_SIZE, 2);
 
     // Then we copy only the needed bytes out of the buffer into our profile.
     size_t bt_bytes = bt_size * sizeof(jl_bt_element_t);
     jl_bt_element_t *bt_data = (jl_bt_element_t*) malloc(bt_bytes);
-    memcpy(bt_data, static_bt_data, bt_bytes);
+    memcpy(bt_data, shared_bt_data_buffer, bt_bytes);
+
+    // Now, "clear" the ptls buffer, so that this buffer isn't incorrectly rooting objects
+    // in that buffer.
+    ptls->bt_size = 0;
```

Comment:
doesn't look like you are rooting this anyways
```suggestion
```

Reply:
Though since it is not being rooted upon return, you need to avoid examining the `jl_pgcstack` in the `rec_backtrace`

Reply:
oops yep you're totally right of course. this is leftover cruft. Thank you!

Reply:
> Though since it is not being rooted upon return, you need to avoid examining the `jl_pgcstack` in the `rec_backtrace`

Can you expand on this? Thanks for your help.

Reply:
`rec_backtrace` assumes the result is guaranteed to always be in a GC rooted buffer, so it can later be safely unpacked and examined. This does not appear to ensure that safety.

Reply:
Hey @vtjnash  — Thanks. You're right that without rooting, some of the pointers recorded here will be unsafe to dereference later.

The problem is, even if we were to fix that problem for this buffer, the backtraces would still be immediately copied to the profiler's vector of allocations, which isn't rooted either: https://github.com/JuliaLang/julia/blob/1ad2396f05fa63a71e5842c814791cd7c7715100/src/gc-alloc-profiler.cpp#L30

So we'd have to fix it there too for these pointers to be safely unpacked and dereferenced.

We spent a bunch of time in https://github.com/JuliaLang/julia/pull/42768 trying to figure out how to root that buffer, before noticing that the CPU profiler just sidesteps this problem — it doesn't root the main buffer, and just renders stack frames it can't dereference as `unknown`. In practice, with the alloc profiler as well as the CPU profiler, there haven't been too many `unknown` frames. So, we ultimately mimicked the CPU profiler.

I filed https://github.com/JuliaLang/julia/issues/43796 to address this for both the CPU and alloc profiler — a bigger project which may involve special-casing the GC to scan both their backtrace buffers, as well as these intermediate buffers. In the meantime, I think 1.8 would be in pretty good shape if we fix this thread safety bug but not fix the few `unknown`s.